src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs

```diff
@@ -304,9 +304,93 @@ public void TestStartOnOSXWithBadFormat()
             }
         }
 
+        [Fact]
+        public void TestStartWithNonExistingUserThrows()
+        {
+            Process p = CreateProcessPortable(RemotelyInvokable.Dummy);
+            p.StartInfo.UserName = "DoesNotExist";
+            Assert.Throws<Win32Exception>(() => p.Start());
+        }
+
+        /// <summary>
+        /// Tests when running as a normal user and starting a new process as the same user
+        /// works as expected.
+        /// </summary>
+        [Fact]
+        public void TestStartWithNormalUser()
+        {
+            TestStartWithUserName();
+        }
+
+        /// <summary>
+        /// Tests when running as root and starting a new process as a normal user,
+        /// the new process doesn't have elevated privileges.
+        /// </summary>
+        [Fact]
+        [OuterLoop("Needs sudo access")]
+        [Trait(XunitConstants.Category, XunitConstants.RequiresElevation)]
+        public void TestStartWithRootUser()
+        {
+            RemoteInvokeOptions options = new RemoteInvokeOptions()
+            {
+                Start = false,
+                RunAsSudo = true
+            };
+            Process p = RemoteInvoke((Func<int>)TestStartWithUserName, options).Process;
+            AddProcessForDispose(p);
+
+            p.Start();
+            p.WaitForExit();
+
+            Assert.Equal(0, p.ExitCode);
+        }
+
+        public static int TestStartWithUserName()
+        {
+            using (ProcessTests testObject = new ProcessTests())
+            {
+                string realUserName;
+                if (geteuid() == 0)
+                {
+                    realUserName = Environment.GetEnvironmentVariable("SUDO_USER");
+                }
+                else
+                {
+                    realUserName = Environment.UserName;
+                }
+
+                Assert.NotNull(realUserName);
+                Assert.NotEqual("root", realUserName);
+
+                using (Process p = testObject.CreateProcessPortable(TestStartWithUserName_ChildProcess))
+                {
+                    p.StartInfo.UserName = realUserName;
+                    Assert.True(p.Start());
+
+                    p.WaitForExit();
+
+                    // since the process was started with the current real user, even if this test
+                    // was run with 'sudo', the child process will be run as the normal real user.
+                    // Assert that the effective user of the child process was never 'root'
+                    // and was the real user of this process.
+                    Assert.NotEqual(0, p.ExitCode);
+                }
+
+                return 0;
```

Comment:
were you intending to always return 0? you are asserting the return code is 0.

Reply:
Yes, there are 2 processes at play here.  There's the outer process that gets invoked with `sudo`.  That's what this exit code is returning.  Then there's the inner/child process that gets invoked as the `realUserName`, which returns its `geteuid()`.

Here I am returning `0` to indicate back to the test that everything worked successfully.